raw-spans-grouper/raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/RawSpansGrouper.java

```diff
@@ -66,48 +60,17 @@ public StreamsBuilder buildTopology(
     Serde valueSerde = defaultValueSerde(properties);
     Serde keySerde = defaultKeySerde(properties);
 
-    long spanWindowStoreRetentionTimeMins =
-        getAppConfig().hasPath(SPAN_WINDOW_STORE_RETENTION_TIME_MINS)
-            ? getAppConfig().getLong(SPAN_WINDOW_STORE_RETENTION_TIME_MINS)
-            : 60;
-    long spanWindowStoreSegmentSizeMins =
-        getAppConfig().hasPath(SPAN_WINDOW_STORE_SEGMENT_SIZE_MINS)
-            ? getAppConfig().getLong(SPAN_WINDOW_STORE_SEGMENT_SIZE_MINS)
-            : 20;
-
-    boolean spanWindowStoreCachingEnabled =
-        !getAppConfig().hasPath(SPAN_WINDOW_STORE_CHACHING_ENABLED)
-            || getAppConfig().getBoolean(SPAN_WINDOW_STORE_CHACHING_ENABLED);
-    StoreBuilder<WindowStore<SpanIdentity, RawSpan>> spanWindowStoreBuilder =
-        Stores.windowStoreBuilder(
-            new RocksDbWindowBytesStoreSupplier(
-                SPAN_WINDOW_STORE,
-                // retention period of window
-                // so data older than 1 hour will be cleaned up
-                Duration.ofMinutes(spanWindowStoreRetentionTimeMins).toMillis(),
-                // length of a segment in rocksdb, so if segment size is 5mins and retention is
-                // 60mins
-                // there will be 12 segments in rocksdb
-                Duration.ofMinutes(spanWindowStoreSegmentSizeMins).toMillis(),
-                // duration of a window,
-                // this param doesn't play any role in actual persistence of data
-                // and is more of a logical construct used while returning the data
-                Duration.ofMinutes(1).toMillis(),
-                false,
-                false),
-            keySerde,
-            valueSerde);
-
-    if (spanWindowStoreCachingEnabled) {
-      spanWindowStoreBuilder.withCachingEnabled();
-    }
-
     StoreBuilder<KeyValueStore<TraceIdentity, TraceState>> traceStateStoreBuilder =
         Stores.keyValueStoreBuilder(
                 Stores.persistentKeyValueStore(TRACE_STATE_STORE), keySerde, valueSerde)
             .withCachingEnabled();
 
-    streamsBuilder.addStateStore(spanWindowStoreBuilder);
+    StoreBuilder<KeyValueStore<SpanIdentity, RawSpan>> spanStoreBuilder =
```

Comment:
For note: I was also thinking of an alernative approach where we fixed the batch. So, we store a batch of fixed size say `10` in the store and create a link with each batch. So, as an example, if the trace is having 100 spans, there are 10 batches of 10 spans that will be stored in k,v store. So, record size at write time (insert) will be limited, and during read time, we will have (number of spans/batch size) queries. So, this way, it can help in reducing punctuate time.

e.g batch size structure.
```
record RawSpansBatch {
List<RawSpan> rawSpans;
TraceIdentity traceIdentity;
}
KeyValueStore<TraceIdentity, RawSpansBatch>
```

No need to do as part of this PR, we can explore it as follow up if it helps in reducing punctuate latency, we can consider.

Reply:
yeah, depending on performance result of testing this approach

Reply:
so what is this change about ? What were the issues with the old approach ?

Reply:
The old approach was not scaling very well due to the use of window store as the state required to be maintained was having a significant cost on the performance and resource consumption.

Reply:
> so what is this change about ? 

are you asking about this particular pr or the approach ronak is suggesting?

Reply:
Curious about the redesign

Reply:
So, the current change is exactly same as the earlier approach only difference is the use of kv store instead of window store. 
This is solving following problems (wrt to the original design):
1. write time is super simple (just an append of span)
2. since a single span is written only once to the store there is no repeated serde cost (n^2 order in the original code were a large blob was getting updated again and again)
3. write to changelog topic won't fail due to message size (earlier as the trace grew it started failing and cpu cycles were wasted in retires)
4. Now the read (punctuate time) is bit heavy, since all the spans have to queried from the kv store 

This pr https://github.com/hypertrace/hypertrace-ingester/pull/193 tries to solve point `4` as well 

Instead of writing a single span to the store, write groups of spans together (say 10)
this will make the write part little heavy and the read (punctuate) a little lightweight 
This is under perf testing as well

Reply:
I feel we should go with #193 as this current design (of writing a single span) can be achieved with it as well. 
Also if we maintain chunks of say 10 spans together write is not going to be costly but our reads at puctuate will decrease by a factor of 10 

I will share some performance numbers soon

Reply:
> I will share some performance numbers soon

The number of times app went into rebalance state with #193 was less compared to this pr, while clearing lag of 20millions span, however the time taken to finish up all the lag was same ~6hrs 

Moving ahead with this pr.